src/main/java/org/opensearch/security/support/StreamableRegistry.java

```diff
@@ -0,0 +1,135 @@
+    }
+
+    private Writeable.Reader<Object> getReader(int id) {
+        if (!idToEntryMap.containsKey(id)) {
+            throw new OpenSearchException(String.format("No reader registered for id %s", id));
+        }
+        return idToEntryMap.get(id).reader;
+    }
+
+    private int getId(Class<?> clazz) {
+        if (!classToIdMap.containsKey(clazz)) {
+            throw new OpenSearchException(String.format("No writer registered for class %s", clazz.getName()));
+        }
+        return classToIdMap.get(clazz);
+    }
+
+    protected boolean isStreamable(Class<?> clazz) {
+        return classToIdMap.containsKey(clazz);
+    }
+
+    protected void writeTo(StreamOutput out, Object object) throws IOException {
+        out.writeByte((byte) getId(object.getClass()));
+        getWriter(object.getClass()).write(out, object);
+    }
+
+    protected Object readFrom(StreamInput in) throws IOException {
+        int id = in.readByte();
+        return getReader(id).read(in);
+    }
+
+    protected static StreamableRegistry getInstance() {
+        return INSTANCE;
+    }
+
+    protected void registerStreamable(Class<?> clazz, Writeable.Writer<Object> writer, Writeable.Reader<Object> reader) {
+        if (Writeable.class.isAssignableFrom(clazz)) {
+            throw new IllegalArgumentException(
+                String.format("%s is Writeable and should not be registered as a streamable", clazz.getName())
+            );
+        }
+        Integer id = classToIdMap.size() + 1;
+        classToIdMap.put(clazz, id);
+        idToEntryMap.put(id, new Entry(writer, reader));
+    }
+
+    protected int getStreamableID(Class<?> clazz) {
+        if (!isStreamable(clazz)) {
+            throw new OpenSearchException(String.format("class %s is in streamable registry", clazz.getName()));
+        } else {
+            return classToIdMap.get(clazz);
+        }
+    }
+
+    /**
+     * Register all streamables here.
+     * <br/>
+     * Caution - Register new streamables towards the end. Removing / reordering a registered streamable will change the typeIDs associated with the streamables
+     * causing a breaking change in the serialization format.
+     */
+    private void registerAllStreamables() {
+
+        // InetSocketAddress
+        this.registerStreamable(InetSocketAddress.class, (Writeable.Writer<Object>) (o, v) -> {
+            final InetSocketAddress inetSocketAddress = (InetSocketAddress) v;
+            o.writeString(inetSocketAddress.getHostString());
+            o.writeByteArray(inetSocketAddress.getAddress().getAddress());
+            o.writeInt(inetSocketAddress.getPort());
+        }, (Writeable.Reader<Object>) (i) -> {
+            String host = i.readString();
+            byte[] addressBytes = i.readByteArray();
+            int port = i.readInt();
+            return new InetSocketAddress(InetAddress.getByAddress(host, addressBytes), port);
+        });
```

Comment:
Me and Peter and a discussion on this here (we also discussed a bit offline) - https://github.com/opensearch-project/security/pull/2802#discussion_r1223254272

We had concluded that the approach is okay since we do not expose any public interface to register Streamables from outside this class. We've added comments and UTs to support this.

Reply:
@parasjain1 - Sorry, I didn't fully understand why it can't be hardcoded? Do you see any cons with it?

Reply:
Hardcoding is possible as well. 

I had just followed the approach that is also followed by other libraries such as Kryo / FST for class pre-registration.

Reply:
Kryo/FST are generic libraries, so they can't be hard coded up front. Given that this is a custom implementation, it makes it easier while debugging if it is hard coded.

Reply:
Resolving as the feedback has been incorporated.